{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "71a6ca26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "from pathlib import Path\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96719d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_arguments():\n",
    "    \"\"\"\n",
    "    Parse the arguments\n",
    "    \"\"\"\n",
    "    parser = argparse.ArgumentParser(description=\"REDD pipelines script generator\",add_help=True)    \n",
    "    \n",
    "    output_group = parser.add_argument_group('Named arguments about outputs')\n",
    "    # output group\n",
    "    output_group.add_argument('-n','--output_name', type=str, help=\"The name of output results and results will be at OUTPUT_PATH/outputs/OUTPUT_NAME.prediction.txt\",required=True)\n",
    "    output_group.add_argument('-o','--output_path', type=str, help=\"The path of output directory\",required=True)\n",
    "    \n",
    "    # input group\n",
    "    input_group = parser.add_argument_group('Named arguments about inputs')\n",
    "    input_group.add_argument('--input_fastq', type=str, help=\"The path of directory containing input fastq files (/fastq_pass/)\",required=True)\n",
    "    input_group.add_argument('--input_fast5', type=str, help=\"The path of directory containing input fast5 files  (/fast5_pass/)\",required=True)\n",
    "    input_group.add_argument('--input_summary', type=str, help=\"The path of sequencing summary file (sequencing_summary_*.txt)\",required=True)\n",
    "    \n",
    "    # reference group\n",
    "    ref_group = parser.add_argument_group('Named arguments about reference files')\n",
    "    ref_group.add_argument('--ref_transcriptome', type=str, help=\"The path of reference transcriptome\",default=None,required=False)\n",
    "    \n",
    "    ref_group.add_argument('--ref_genome', type=str, help=\"The path of reference genome\",required=True)\n",
    "    ref_group.add_argument('--ref_cdna_to_genome', type=str, help=\"The path of reference cdna to genome file\",default=None,required=False)\n",
    "    ref_group.add_argument('--ref_annotation', type=str, help=\"The path of reference annotation\",default=None,required=False)\n",
    "    ref_group.add_argument('--ref_alu', type=str, help=\"The path of reference alu\",required=True)\n",
    "    ref_group.add_argument('--ref_snp', type=str, help=\"The path of reference snp\",default=None,required=False)\n",
    "    ref_group.add_argument('--ref_m6A_motif', type=str, help=\"The path of reference m6A motif\",default=None,required=False)\n",
    "    ref_group.add_argument('--ref_REDIportal', type=str, help=\"The path of reference REDI portal file\",required=True)\n",
    "    \n",
    "    # pipeline groups\n",
    "    pipeline_group = parser.add_argument_group('Named arguments about other pipeline settings')\n",
    "    pipeline_group.add_argument('--overall_time', type=int, help=\"The estimated overall running time\",default=24,required=False)\n",
    "    pipeline_group.add_argument('--pipeline_mode', type=str, help=\"Which mode to use to run the pipeline.[cluster,bash]\",default='cluster',required=False)\n",
    "    pipeline_group.add_argument('--account', type=str, help=\"Which account uses for running in cluster. Required if --pipeline_mode set to cluster\",default=None,required=False)\n",
    "    pipeline_group.add_argument('--max_cores_resources', type=int, help=\"Max number of cores uses in the same time.\",default=112,required=False)\n",
    "    pipeline_group.add_argument('--num_split', type=int, help=\"Number of file split to enable parallel computing\",default=4,required=False)\n",
    "    pipeline_group.add_argument('--reference_choice',type=str,help=\"Which reference will be used [genome,transcriptome]\",default='genome',required=False)\n",
    "    pipeline_group.add_argument('--model',type=str,help=\"Which model will be used [general]\",default='general',required=False)\n",
    "    pipeline_group.add_argument('--filter_snp',type=str,help=\"Whether filer out snp\",default='True',required=False)\n",
    "    pipeline_group.add_argument('--filter_m6A',type=str,help=\"Whether filer out m6A\",default='True',required=False)\n",
    "    pipeline_group.add_argument('--device',type=str,help=\"Use GPU or CPU to speed up prediction [CPU,GPU]\",default='CPU',required=False)\n",
    "    args = parser.parse_args()\n",
    "    return args\n",
    "def prompt_error(error_str):\n",
    "    print(error_str)\n",
    "    exit()\n",
    "\n",
    "\n",
    "def main():\n",
    "    args = parse_arguments()\n",
    "    \n",
    "    output_folder = args.output_path\n",
    "    output_name = args.output_name\n",
    "    \n",
    "    input_fastq_folder = args.input_fastq\n",
    "    input_fast5_folder = args.input_fast5\n",
    "    input_summary_file = args.input_summary\n",
    "    \n",
    "    ref_transcriptome_file = args.ref_transcriptome\n",
    "    ref_genome_file = args.ref_genome\n",
    "    ref_cdna_to_genome_file = args.ref_cdna_to_genome\n",
    "    ref_annotation_file = args.ref_annotation\n",
    "    ref_alu_file = args.ref_alu_file\n",
    "    ref_snp_file = args.ref_snp_file\n",
    "    ref_m6A_motif_file = args.ref_m6A_motif\n",
    "    ref_REDIportal_file = args.ref_REDIportal\n",
    "    \n",
    "    overall_time = args.overall_time\n",
    "    accout = args.account\n",
    "    max_cores_resources = args.max_cores_resources\n",
    "    num_split = args.num_split\n",
    "    reference_choice = args.reference_choice\n",
    "    model = args.model\n",
    "    filter_snp = args.filter_snp\n",
    "    filter_m6A = args.filter_m6A\n",
    "    device = args.device\n",
    "    pipeline_mode = args.pipeline_mode\n",
    "    Path(output_folder).mkdir(exist_ok=True,parents=True)\n",
    "    # error handling\n",
    "    \n",
    "    if filter_snp == 'True' and ref_snp_file is None:\n",
    "        prompt_error('No reference snp file(--ref_snp) is given but you have set --filter_snp to True')\n",
    "    if filter_m6A == 'True' and ref_m6A_file is None:\n",
    "        prompt_error('No reference m6A file(--ref_m6A) is given but you have set --filter_m6A to True')\n",
    "    \n",
    "    if pipeline_mode == 'cluster' and account is None:\n",
    "        prompt_error('No account(--account) is given but you have set --pipeline_mode to cluster')\n",
    "    if reference_choice == 'genome':\n",
    "        with open(f'{output_folder}/empty', 'w') as fp:\n",
    "            pass\n",
    "        os.system(f\"attrib +h {output_folder}/empty\")\n",
    "        ref_transcriptome_file = f'{output_folder}/empty'\n",
    "        ref_cdna_to_genome_file = f'{output_folder}/empty'\n",
    "        ref_annotation_file = f'{output_folder}/empty'\n",
    "        ref_dump_position = 'disk'\n",
    "    else:\n",
    "        if ref_transcriptome_file is None:\n",
    "            prompt_error('No reference transcriptome(--ref_transcriptome)  is given but you have set --reference_choice to transcriptome')\n",
    "        if ref_cdna_to_genome_file is None:\n",
    "            prompt_error('No reference cdna_to_genome_file(--ref_cdna_to_genome)  is given but you have set --reference_choice to transcriptome')\n",
    "        if ref_annotation_file is None:\n",
    "            prompt_error('No reference annotation(--ref_annotation) is given but you have set --reference_choice to transcriptome')\n",
    "        ref_dump_position = 'memory'\n",
    "    ref_snp_file = f'{output_folder}/empty' if ref_alu_file is None\n",
    "    ref_m6A_motif_file = f'{output_folder}/empty' if ref_m6A_motif_file is None\n",
    "        \n",
    "    config = f'''\n",
    "reference: '{reference_choice}' # genome or transcriptome\n",
    "center: 'A'\n",
    "nt: 4\n",
    "featurenum: 5\n",
    "buffersize: 1000\n",
    "labeltype: 'I'\n",
    "ref_dump_position: \"{ref_dump_position}\"\n",
    "device: '{device}'\n",
    "batch_size: 10000\n",
    "model: '{model}'\n",
    "coverage_cutoff: 5\n",
    "ratio_cutoff: 0\n",
    "filter_snp: '{filter_snp}'\n",
    "filter_m6A: '{filter_m6A}'\n",
    "'''\n",
    "    \n",
    "    script = f'''#!/bin/bash\n",
    "#SBATCH --time={overall_time}:00:00\n",
    "#SBATCH --nodes=1 --ntasks-per-node=1\n",
    "#SBATCH --account={account}\n",
    "\n",
    "source activate AIediting\n",
    "\n",
    "mkdir {output_folder}/intermediates\n",
    "mkdir {output_folder}/intermediates/fastq/\n",
    "mkdir {output_folder}/intermediates/fast5/\n",
    "mkdir {output_folder}/intermediates/summary/\n",
    "mkdir {output_folder}/intermediates/reference/\n",
    "# link data\n",
    "ln -s {input_fastq_folder} {output_folder}/intermediates/fastq/{output_name}\n",
    "ln -s {input_fast5_folder} {output_folder}/intermediates/fast5/{output_name}\n",
    "ln -s {input_summary_file} {output_folder}/intermediates/summary/{output_name}_summary.txt\n",
    "# link reference\n",
    "ln -s {ref_transcriptome_file} {output_folder}/intermediates/reference/transcriptome.fa\n",
    "ln -s {ref_genome_file} {output_folder}/intermediates/reference/genome.fa\n",
    "ln -s {ref_cdna_to_genome_file} {output_folder}/intermediates/reference/gencode.v31.annotation.cdna2genome.tab\n",
    "ln -s {ref_annotation_file} {output_folder}/intermediates/reference/gencode.v31.annotation.gpd\n",
    "ln -s {ref_alu_file} {output_folder}/intermediates/reference/Hg38_Alu.merge.bed\n",
    "ln -s {ref_snp_file} {output_folder}/intermediates/reference/hg38_snp151.bed\n",
    "ln -s {ref_m6A_motif_file} {output_folder}/intermediates/reference/analysis_extract_m6A_motif.bed\n",
    "ln -s {ref_REDIportal_file} {output_folder}/intermediates/reference/REDIportal_hg38.txt\n",
    "\n",
    "# link script\n",
    "ln -s {snakefile_path}/scripts scripts\n",
    "ln -s {snakefile_path}/Snakefile Snakefile\n",
    "\n",
    "snakemake --unlock\n",
    "snakemake -c {max_cores_resources} -p outputs/precomputed_visualization/{output_name} --cluster \"sbatch -A {account} -t {{resources.runtime}} -N 1 -c {{threads}}\" --jobs {num_split} --set-scatter split={num_split} --latency-wait 60 --rerun-incomplete\n",
    "'''\n",
    "    with open(f'{output_folder}/run.pbs','w') as f:\n",
    "        f.write(script)\n",
    "    with open(f'{output_folder}/config.yaml','w') as f:\n",
    "        f.write(config)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f0ff22ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output_name = 'H1-DE_genome'\n",
    "# overall_time = 24\n",
    "# account = 'PAS1669'\n",
    "# max_cores_resources = 112\n",
    "# num_split = 4\n",
    "# reference_choice = 'genome'\n",
    "# output_folder = '/fs/ess/scratch/PCON0009/haoran/test_AIediting'\n",
    "# input_fastq_folder = '/fs/project/PCON0009/LabData/Human/DE-H1/Transcriptome/ONT/directRNA/20210330_2124_X2_FAP47598_1c046625/fastq_pass'\n",
    "# input_fast5_folder = '/fs/project/PCON0009/LabData/Human/DE-H1/Transcriptome/ONT/directRNA/20210330_2124_X2_FAP47598_1c046625/fast5_pass'\n",
    "# input_summary_file = '/fs/project/PCON0009/LabData/Human/DE-H1/Transcriptome/ONT/directRNA/20210330_2124_X2_FAP47598_1c046625/sequencing_summary_FAP47598_07e34f33.txt'\n",
    "# ref_transcriptome_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/2-curate_gtf/Stem_cell_talon.flt.bam_flt.gtf.fa'\n",
    "# ref_genome_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/ref/genome.fa'\n",
    "# ref_cdna_to_genome_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/gencode.v31.annotation.cdna2genome.tab'\n",
    "# ref_annotation_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/gencode.v31.annotation.gpd'\n",
    "# ref_alu_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/Hg38_Alu.merge.bed'\n",
    "# ref_snp_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/hg38_snp151.bed'\n",
    "# ref_m6A_motif_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/analysis_extract_m6A_motif.bed'\n",
    "# ref_REDIportal_file = '/fs/project/PCON0009/Au-scratch2/ying/StemCell/RNAdirect/reditools2/test/REDIportal_hg38.txt'\n",
    "# snakefile_path = '/users/PCON0009/haoranli/_projects/AIediting_pipelines/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "650569b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a160987",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "831082ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ca19b605",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eedb0ec2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e3fdcfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "67430539",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/users/PCON0009/haoranli/_projects/AIediting_pipelines/outputs/H1-DE_FAP47598_1c046625.site.bed',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1a4caa8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENST00000000233.10</th>\n",
       "      <th>19</th>\n",
       "      <th>3</th>\n",
       "      <th>75</th>\n",
       "      <th>0.040</th>\n",
       "      <th>0.037</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>43</td>\n",
       "      <td>1</td>\n",
       "      <td>121</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>132</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>98</td>\n",
       "      <td>3</td>\n",
       "      <td>189</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7800410</th>\n",
       "      <td>TALONT002310543</td>\n",
       "      <td>1678</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7800411</th>\n",
       "      <td>TALONT002310543</td>\n",
       "      <td>1679</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7800412</th>\n",
       "      <td>TALONT002310543</td>\n",
       "      <td>1680</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7800413</th>\n",
       "      <td>TALONT002310543</td>\n",
       "      <td>1681</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7800414</th>\n",
       "      <td>TALONT002310543</td>\n",
       "      <td>1684</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7800415 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ENST00000000233.10    19  3   75  0.040  0.037\n",
       "0        ENST00000000233.10    37  1  104  0.010  0.028\n",
       "1        ENST00000000233.10    43  1  121  0.008  0.015\n",
       "2        ENST00000000233.10    50  1  132  0.008  0.026\n",
       "3        ENST00000000233.10    89  0  190  0.000  0.002\n",
       "4        ENST00000000233.10    98  3  189  0.016  0.023\n",
       "...                     ...   ... ..  ...    ...    ...\n",
       "7800410     TALONT002310543  1678  0   20  0.000  0.000\n",
       "7800411     TALONT002310543  1679  0   20  0.000  0.001\n",
       "7800412     TALONT002310543  1680  0   19  0.000  0.002\n",
       "7800413     TALONT002310543  1681  0   19  0.000  0.000\n",
       "7800414     TALONT002310543  1684  0   15  0.000  0.002\n",
       "\n",
       "[7800415 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "469aedb0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupnb",
   "language": "python",
   "name": "jupnb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
